src/fu-self-test.c

```diff
@@ -75,9 +76,49 @@ fu_device_metadata_func (void)
 }
 
 static void
+fu_smbios_func (void)
+{
+	const gchar *str;
+	gboolean ret;
+	g_autofree gchar *dump = NULL;
+	g_autofree gchar *sysfsdir = NULL;
+	g_autoptr(FuSmbios) smbios = NULL;
+	g_autoptr(GError) error = NULL;
+
+	sysfsdir = fu_test_get_filename (TESTDATADIR, ".");
+	g_assert (sysfsdir != NULL);
+
+	smbios = fu_smbios_new ();
+	ret = fu_smbios_setup (smbios, sysfsdir, &error);
+	g_assert_no_error (error);
+	g_assert (ret);
+	dump = fu_smbios_to_string (smbios);
+	if (g_getenv ("VERBOSE") != NULL)
+		g_debug ("%s", dump);
+
+	/* test for missing table */
+	str = fu_smbios_get_string (smbios, 0xff, 0, &error);
+	g_assert_error (error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE);
+	g_assert_null (str);
+	g_clear_error (&error);
+
+	/* check for invalid offset */
+	str = fu_smbios_get_string (smbios, FU_SMBIOS_STRUCTURE_TYPE_BIOS, 0xff, &error);
+	g_assert_error (error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE);
+	g_assert_null (str);
+	g_clear_error (&error);
+
+	/* get vendor */
+	str = fu_smbios_get_string (smbios, FU_SMBIOS_STRUCTURE_TYPE_BIOS, 0x04, &error);
```

Comment:
It would probably add some defines for what `0x04` really is elsewhere than magic number.

Reply:
So, I pondered that as well. It's a tricky one as I don't want to have thousands of #defines basically duplicating the document here: https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.1.1.pdf

Realistically, if you're using the `fu_smbios_` functions you need to have a copy of that specification document open. I'll add a note to that specification in the code.

Reply:
OK.